contracts/rate_oracles/CompoundRateOracle.sol

```diff
@@ -0,0 +1,255 @@
+// SPDX-License-Identifier: Apache-2.0
+
+pragma solidity =0.8.9;
+
+import "../interfaces/rate_oracles/ICompoundRateOracle.sol";
+import "../interfaces/compound/ICToken.sol";
+import "../core_libraries/FixedAndVariableMath.sol";
+import "../utils/WadRayMath.sol";
+import "../rate_oracles/BaseRateOracle.sol";
+
+contract CompoundRateOracle is BaseRateOracle, ICompoundRateOracle {
+    using OracleBuffer for OracleBuffer.Observation[65535];
+
+    /// @inheritdoc ICompoundRateOracle
+    ICToken public override ctoken;
+
+    /// @inheritdoc ICompoundRateOracle
+    uint256 public override decimals;
+
+    uint8 public constant override UNDERLYING_YIELD_BEARING_PROTOCOL_ID = 2; // id of compound is 2
+
+    constructor(
+        ICToken _ctoken,
+        IERC20Minimal underlying,
+        uint8 _decimals
+    ) BaseRateOracle(underlying) {
+        ctoken = _ctoken;
+        require(
+            ctoken.underlying() == address(underlying),
+            "Tokens do not match"
+        );
+        decimals = _decimals;
+        uint32 blockTimestamp = Time.blockTimestampTruncated();
+        uint256 result = exchangeRateInRay();
+        (
+            oracleVars.rateCardinality,
+            oracleVars.rateCardinalityNext
+        ) = observations.initialize(blockTimestamp, result);
+    }
+
+    function exchangeRateInRay() internal view returns (uint256 resultRay) {
+        // cToken exchangeRateStored() returns the current exchange rate as an unsigned integer, scaled by 1 * 10^(10 + Underlying Token Decimals)
+        // source: https://compound.finance/docs/ctokens#exchange-rate and https://compound.finance/docs#protocol-math
+        // We want the same number scaled by 10^27 (ray)
+        // So: if Underlying Token Decimals == 17, no scaling is required
+        //     if Underlying Token Decimals > 17, we scale down by a factor of 10^difference
+        //     if Underlying Token Decimals < 17, we scale up by a factor of 10^difference
+        // And if decimals
```

Comment:
don't get this comment

Reply:
Have deleted the "And if decimals" line. Do you get the rest of it?

Reply:
yeah, the rest makes sense